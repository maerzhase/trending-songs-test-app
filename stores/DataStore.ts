import { makeObservable, observable, action, toJS } from 'mobx'
import { getSongsTrending, postSongLike, postSongUnlike } from '../api/index'
import Song from './models/Song'

export interface IDataStore {
  songs: Song[]
  initialize(): Promise<Song[]>
  receiveSongs(songs: Song[]): void
  likeSong(song: Song): Promise<boolean>
  unlikeSong(song: Song): Promise<boolean>
  getSnapshot(): IDataStoreSnapshot
  hydrate(snapshot: IDataStoreSnapshot): void
}

interface IDataStoreSnapshot {
  songs: Song[]
}

class DataStore implements IDataStore {
  constructor() {
    makeObservable(this, {
      songs: observable,
      initialize: action,
      receiveSongs: action,
      likeSong: action,
      unlikeSong: action,
    })
  }

  public songs: Song[] = []

  public initialize = async (): Promise<Song[]> => {
    try {
      const songs = await getSongsTrending()
      this.receiveSongs(songs)
      return songs
    } catch (e) {
      console.warn(e)
    }
  }

  public receiveSongs = (songs: Song[]): void => {
    this.songs = songs
  }

  public likeSong = async (song: Song): Promise<boolean> => {
    try {
      await postSongLike(song)
      await this.initialize()
      return true
    } catch (e) {
      console.warn(e)
    }
  }

  public unlikeSong = async (song: Song): Promise<boolean> => {
    try {
      await postSongUnlike(song)
      await this.initialize()
      return true
    } catch (e) {
      console.warn(e)
    }
  }

  public getSnapshot = (): IDataStoreSnapshot => {
    const { hydrate, getSnapshot, ...dataStoreSnapshot } = this // eslint-disable-line
    return toJS(dataStoreSnapshot)
  }

  public hydrate = (snapshot: IDataStoreSnapshot): void => {
    if (snapshot.songs) this.songs = snapshot.songs
  }
}

export default DataStore
