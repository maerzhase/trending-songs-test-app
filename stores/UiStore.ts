import { makeObservable, observable, action } from 'mobx'

export interface IUiStore {
  activeSongId?: string
  setActiveSongId(id: string): void
}

class UiStore implements IUiStore {
  constructor() {
    makeObservable(this, {
      activeSongId: observable,
      setActiveSongId: action,
    })
  }

  public activeSongId?: string = null

  public setActiveSongId = (id: string): void => {
    this.activeSongId = id
  }
}

export default UiStore
