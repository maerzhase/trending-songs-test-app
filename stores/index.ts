import React, { useMemo } from 'react'
import DataStore, { IDataStore } from './DataStore'
import UiStore, { IUiStore } from './UiStore'
import { enableStaticRendering } from 'mobx-react'
import { MobXProviderContext } from 'mobx-react'

enableStaticRendering(typeof window === 'undefined')

declare global {
  interface Window {
    __ds: IDataStore
    __us: IUiStore
  }
}

interface IInitialState {
  dataStore: IDataStore
  uiStore: IUiStore
}

let ds, us

export const getStoreInstances = (
  initialState: IInitialState
): IInitialState => {
  const { dataStore } = initialState || {}
  const _dataStore = ds ?? new DataStore()
  const _uiStore = us ?? new UiStore()

  // _dataStore.connectStores({ uiStore: _uiStore });

  if (dataStore) _dataStore.hydrate(dataStore)

  // For SSG and SSR always create a new store
  if (typeof window === 'undefined')
    return { dataStore: _dataStore, uiStore: _uiStore }

  // Create the stores once in the client
  if (!ds) ds = _dataStore
  if (!us) us = _uiStore

  if (global.window) {
    window.__ds = ds
    window.__us = us
  }

  return { dataStore: ds, uiStore: us }
}

export function useStoresForProviders(
  initialState: IInitialState
): IInitialState {
  const stores = useMemo(() => getStoreInstances(initialState), [initialState])
  return stores
}

const useStores = (): Record<string, any> => {
  return React.useContext(MobXProviderContext)
}

export default useStores
