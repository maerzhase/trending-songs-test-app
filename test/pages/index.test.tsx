import React from 'react'
import { render } from '../testUtils'
import { getStoreInstances } from '../../stores/index'
import Home from '../../pages/index'

const { dataStore } = getStoreInstances(null)

const initStores = async () => {
  await dataStore.initialize()
  return {}
}

describe('Home page', () => {
  it('matches snapshot', () => {
    return initStores().then(() => {
      const { asFragment } = render(<Home />, {})
      expect(asFragment()).toMatchSnapshot()
    })
  })

  // it('clicking button triggers alert', () => {
  //   const { getByText } = render(<Home />, {})
  //   window.alert = jest.fn()
  //   fireEvent.click(getByText('Test Button'))
  //   expect(window.alert).toHaveBeenCalledWith('With typescript and Jest')
  // })
})
