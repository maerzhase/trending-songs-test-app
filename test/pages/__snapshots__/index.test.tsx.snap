// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <main>
      <div
        class="MuiBox-root MuiBox-root-1"
      >
        <div
          class="MuiBox-root MuiBox-root-2"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
          >
            <div
              class="MuiBox-root MuiBox-root-3"
            >
              <img
                height="200"
                src="https://dz2l6nhikl0ua.cloudfront.net/assets/cover_image/4801376e-cd1c-11eb-8a0d-022ed69f96e6.jpeg?cb=20210614162451"
                width="200"
              />
              <h6
                class="MuiTypography-root MuiTypography-h6"
              >
                2 Step 1
              </h6>
              <p
                class="MuiTypography-root MuiTypography-body1"
              >
                LF
              </p>
              <div>
                <audio>
                  <source
                    src="https://dz2l6nhikl0ua.cloudfront.net/assets/music_file/44f443f3-cd1c-11eb-8a0d-022ed69f96e6.m4a?cb=20210614162446"
                  />
                  Your browser does not support the audio element.
                </audio>
                <div
                  class="MuiBox-root MuiBox-root-4"
                >
                  <button
                    class="MuiButtonBase-root MuiIconButton-root"
                    tabindex="0"
                    type="button"
                  >
                    <span
                      class="MuiIconButton-label"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M8 5v14l11-7z"
                        />
                      </svg>
                    </span>
                    <span
                      class="MuiTouchRipple-root"
                    />
                  </button>
                  <input
                    class="progress"
                    max="55"
                    min="0"
                    step="1"
                    type="range"
                    value="0"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiBox-root MuiBox-root-5"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
          >
            <div
              class="MuiBox-root MuiBox-root-6"
            >
              <img
                height="200"
                src="https://api-stg.jam-community.com/img/song_default_image.png"
                width="200"
              />
              <h6
                class="MuiTypography-root MuiTypography-h6"
              >
                Lose Control
              </h6>
              <p
                class="MuiTypography-root MuiTypography-body1"
              >
                *                      MrNobody ®
              </p>
            </div>
          </div>
        </div>
        <div
          class="MuiBox-root MuiBox-root-7"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
          >
            <div
              class="MuiBox-root MuiBox-root-8"
            >
              <img
                height="200"
                src="https://api-stg.jam-community.com/img/song_default_image.png"
                width="200"
              />
              <h6
                class="MuiTypography-root MuiTypography-h6"
              >
                Disconnected World
              </h6>
              <p
                class="MuiTypography-root MuiTypography-body1"
              >
                MRNOBODY®
              </p>
            </div>
          </div>
        </div>
        <div
          class="MuiBox-root MuiBox-root-9"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
          >
            <div
              class="MuiBox-root MuiBox-root-10"
            >
              <img
                height="200"
                src="https://dz2l6nhikl0ua.cloudfront.net/assets/cover_image/3f41907e-71f0-11eb-8a0d-022ed69f96e6.jpeg?cb=20210218145023"
                width="200"
              />
              <h6
                class="MuiTypography-root MuiTypography-h6"
              >
                mix
              </h6>
              <p
                class="MuiTypography-root MuiTypography-body1"
              >
                aarelxuxmn
              </p>
              <div>
                <audio>
                  <source
                    src="https://dz2l6nhikl0ua.cloudfront.net/assets/music_file/3f28ee9e-71f0-11eb-8a0d-022ed69f96e6.m4a?cb=20210218145023"
                  />
                  Your browser does not support the audio element.
                </audio>
                <div
                  class="MuiBox-root MuiBox-root-11"
                >
                  <button
                    class="MuiButtonBase-root MuiIconButton-root"
                    tabindex="0"
                    type="button"
                  >
                    <span
                      class="MuiIconButton-label"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M8 5v14l11-7z"
                        />
                      </svg>
                    </span>
                    <span
                      class="MuiTouchRipple-root"
                    />
                  </button>
                  <input
                    class="progress"
                    max="33"
                    min="0"
                    step="1"
                    type="range"
                    value="0"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiBox-root MuiBox-root-12"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
          >
            <div
              class="MuiBox-root MuiBox-root-13"
            >
              <img
                height="200"
                src="https://api-stg.jam-community.com/img/song_default_image.png"
                width="200"
              />
              <h6
                class="MuiTypography-root MuiTypography-h6"
              >
                Luv Me
              </h6>
              <p
                class="MuiTypography-root MuiTypography-body1"
              >
                MRNOBODY®
              </p>
            </div>
          </div>
        </div>
        <div
          class="MuiBox-root MuiBox-root-14"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
          >
            <div
              class="MuiBox-root MuiBox-root-15"
            >
              <img
                height="200"
                src="https://api-stg.jam-community.com/img/song_default_image.png"
                width="200"
              />
              <h6
                class="MuiTypography-root MuiTypography-h6"
              >
                Time Traveller
              </h6>
              <p
                class="MuiTypography-root MuiTypography-body1"
              >
                MRNOBODY®
              </p>
            </div>
          </div>
        </div>
        <div
          class="MuiBox-root MuiBox-root-16"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
          >
            <div
              class="MuiBox-root MuiBox-root-17"
            >
              <img
                height="200"
                src="https://api-stg.jam-community.com/img/song_default_image.png"
                width="200"
              />
              <h6
                class="MuiTypography-root MuiTypography-h6"
              >
                Lean On Me
              </h6>
              <p
                class="MuiTypography-root MuiTypography-body1"
              >
                MRNOBODY®
              </p>
            </div>
          </div>
        </div>
        <div
          class="MuiBox-root MuiBox-root-18"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
          >
            <div
              class="MuiBox-root MuiBox-root-19"
            >
              <img
                height="200"
                src="https://api-stg.jam-community.com/img/song_default_image.png"
                width="200"
              />
              <h6
                class="MuiTypography-root MuiTypography-h6"
              >
                Wonderland
              </h6>
              <p
                class="MuiTypography-root MuiTypography-body1"
              >
                MRNOBODY®
              </p>
            </div>
          </div>
        </div>
        <div
          class="MuiBox-root MuiBox-root-20"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
          >
            <div
              class="MuiBox-root MuiBox-root-21"
            >
              <img
                height="200"
                src="https://api-stg.jam-community.com/img/song_default_image.png"
                width="200"
              />
              <h6
                class="MuiTypography-root MuiTypography-h6"
              >
                World Anthems
              </h6>
              <p
                class="MuiTypography-root MuiTypography-body1"
              >
                *                            MrNobody ®
              </p>
            </div>
          </div>
        </div>
        <div
          class="MuiBox-root MuiBox-root-22"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
          >
            <div
              class="MuiBox-root MuiBox-root-23"
            >
              <img
                height="200"
                src="https://api-stg.jam-community.com/img/song_default_image.png"
                width="200"
              />
              <h6
                class="MuiTypography-root MuiTypography-h6"
              >
                Close Tonight
              </h6>
              <p
                class="MuiTypography-root MuiTypography-body1"
              >
                MRNOBODY®
              </p>
            </div>
          </div>
        </div>
      </div>
    </main>
    <footer />
  </div>
</DocumentFragment>
`;
